File: splitstep/score.py
"""Tennis scoring as a pure function of who won each point.

Only the winner of each point is stored (rallies.winner); everything a
scoreboard shows is replayed from that. Nothing is cached, so a corrected
winner, an undo, a re-segment or a split recomputes for free. The same
function exists in web/src/lib/score.ts for the queue, and both are pinned
to tests/fixtures/score_cases.json -- a case that passes on one side and
fails on the other is the drift that file exists to catch. Change the
rules here and there together, and add a case.
"""

from collections.abc import Mapping, Sequence
from dataclasses import dataclass

_POINT_LABELS = ("0", "15", "30", "40")


@dataclass(frozen=True)
class ScoreRules:
    players: tuple[str, str]
    sets: int
    ad: bool
    tiebreak: str
    tiebreak_to: int


@dataclass(frozen=True)
class ScoreState:
    sets: tuple[tuple[int, int], ...]
    games: tuple[int, int]
    points: tuple[str, str]
    in_tiebreak: bool
    finished: str | None


DEFAULT_RULES = ScoreRules(("Me", "Opp"), 3, True, "at6", 7)


def rules_from_dict(d: Mapping) -> ScoreRules:
    """The JSON column's shape -> rules, validated. Raises ValueError rather
    than returning a default: a session whose rules do not parse must not
    silently score as best-of-three."""
    players = tuple(str(p).strip() for p in d.get("players", ()))
    if len(players) != 2 or not all(players):
        raise ValueError("players must be two non-empty names")
    sets = d.get("sets")
    if sets not in (1, 3, 5):
        raise ValueError("sets must be 1, 3 or 5")
    ad = d.get("ad")
    if not isinstance(ad, bool):
        # ValueError, not TypeError: this validates the shape of a JSON blob
        # from the db or an API body, not a caller's argument type, and the
        # other three checks below raise ValueError for the same reason --
        # a uniform except ValueError at the boundary must catch all of them.
        raise ValueError("ad must be true or false")  # noqa: TRY004
    tiebreak = d.get("tiebreak")
    if tiebreak not in ("at6", "none", "only"):
        raise ValueError("tiebreak must be at6, none or only")
    tiebreak_to = d.get("tiebreakTo")
    if tiebreak_to not in (7, 10):
        raise ValueError("tiebreakTo must be 7 or 10")
    return ScoreRules(players, sets, ad, tiebreak, tiebreak_to)


def rules_to_dict(r: ScoreRules) -> dict:
    return {
        "players": list(r.players),
        "sets": r.sets,
        "ad": r.ad,
        "tiebreak": r.tiebreak,
        "tiebreakTo": r.tiebreak_to,
    }


def _point_labels(pts: list[int], in_tiebreak: bool, ad: bool) -> tuple[str, str]:
    if in_tiebreak:
        return (str(pts[0]), str(pts[1]))
    a, b = pts
    if ad and a >= 3 and b >= 3:
        if a == b:
            return ("40", "40")
        return ("Ad", "40") if a > b else ("40", "Ad")
    return (_POINT_LABELS[min(a, 3)], _POINT_LABELS[min(b, 3)])


def score(winners: Sequence[str], rules: ScoreRules) -> ScoreState:
    """Replay `winners` ('a' / 'b' per point) into a scoreboard state.

    Points after the match is decided are ignored rather than rejected: a
    reviewer who keeps scoring warm-down rallies has made no error the
    scoreboard needs to shout about, and the queue shows them as unscored.
    """
    sets_needed = rules.sets // 2 + 1
    sets: list[tuple[int, int]] = []
    games = [0, 0]
    pts = [0, 0]
    in_tb = rules.tiebreak == "only"
    finished: str | None = None

    for w in winners:
        if finished is not None:
            break
        if w not in ("a", "b"):
            raise ValueError(f"winner must be 'a' or 'b', not {w!r}")
        i = 0 if w == "a" else 1
        j = 1 - i
        pts[i] += 1

        if in_tb:
            if pts[i] >= rules.tiebreak_to and pts[i] - pts[j] >= 2:
                if rules.tiebreak == "only":
                    # The whole session is one tiebreak: the count stays on
                    # the board as the final score, nothing rolls into games.
                    finished = w
                else:
                    games[i] += 1
                    sets.append((games[0], games[1]))
                    games = [0, 0]
                    pts = [0, 0]
                    in_tb = False
                    if sum(1 for s in sets if s[i] > s[j]) >= sets_needed:
                        finished = w
            continue

        # A normal game. With advantage scoring you need four points and two
        # clear; without it the seventh point of a game (4 for the winner)
        # decides it at deuce.
        won_game = pts[i] >= 4 and (not rules.ad or pts[i] - pts[j] >= 2)
        if not won_game:
            continue
        games[i] += 1
        pts = [0, 0]
        if rules.tiebreak == "at6" and games == [6, 6]:
            in_tb = True
        elif games[i] >= 6 and games[i] - games[j] >= 2:
            sets.append((games[0], games[1]))
            games = [0, 0]
            if sum(1 for s in sets if s[i] > s[j]) >= sets_needed:
                finished = w

    return ScoreState(
        sets=tuple(sets),
        games=(games[0], games[1]),
        points=_point_labels(pts, in_tb, rules.ad),
        in_tiebreak=in_tb,
        finished=finished,
    )


def score_before(
    rallies: Sequence[Mapping], rally_id: str, rules: ScoreRules
) -> tuple[ScoreState, int]:
    """The state entering `rally_id`, and how many earlier points carry no
    winner. Replays every non-rejected point with idx below the target's,
    in idx order regardless of the order `rallies` arrived in. An id no row
    holds (an orphaned reel item) replays everything -- the board on such a
    clip is the best information there is, and never a crash."""
    target = next((r for r in rallies if r["id"] == rally_id), None)
    limit = target["idx"] if target is not None else None
    earlier = sorted(
        (r for r in rallies
         if (limit is None or r["idx"] < limit) and not r["rejected"] and r["point"]),
        key=lambda r: r["idx"],
    )
    winners = [r["winner"] for r in earlier if r["winner"]]
    unscored = sum(1 for r in earlier if not r["winner"])
    return score(winners, rules), unscored


def scoreboard_rows(state: ScoreState, rules: ScoreRules) -> list[list[str]]:
    """Two rows for a broadcast-style board: name, one column per completed
    set, current games, points. A tiebreak-only session has no games column
    -- the tiebreak count is the whole score. A finished match replaces the
    games and points columns with a single W for the winner, blank for the
    other; a finished tiebreak-only session keeps its final count instead."""
    rows = []
    for i, name in enumerate(rules.players):
        row = [name, *(str(s[i]) for s in state.sets)]
        if state.finished is not None and rules.tiebreak != "only":
            # Games and points are both 0-0 after the deciding set: showing
            # them would read as a match still in play. Sets, then W.
            row.append("W" if state.finished == ("a", "b")[i] else "")
        else:
            if rules.tiebreak != "only":
                row.append(str(state.games[i]))
            row.append(state.points[i])
        rows.append(row)
    return rows

File: web/src/lib/score.ts
import type { Rally } from './types'

/**
 * Tennis scoring as a pure function of who won each point.
 *
 * Only the winner is stored per rally; the scoreboard is replayed from
 * that, so a corrected winner, an undo, a re-segment or a split recomputes
 * for free. This is the second copy of splitstep/score.py -- the queue
 * needs the score at keypress time, the reel job needs it in Python --
 * and both are pinned to tests/fixtures/score_cases.json. Change the rules
 * in both places together, and add a case.
 */
export type Player = 'a' | 'b'

export interface ScoreRules {
  players: [string, string]
  sets: 1 | 3 | 5
  ad: boolean
  tiebreak: 'at6' | 'none' | 'only'
  tiebreakTo: 7 | 10
  /** Who served the session's first point, or null when nobody said.
   *  Never defaulted: every session recorded before this key existed has
   *  no honest value, and 'a' would be wrong half the time. Absent renders
   *  no server at all, the way an unassigned preset renders no region. */
  firstServer: Player | null
}

export interface ScoreState {
  sets: [number, number][]
  games: [number, number]
  points: [string, string]
  inTiebreak: boolean
  finished: Player | null
  /** Who serves the point this state is entering, derived from
   *  rules.firstServer by the same replay and stored nowhere. Null when no
   *  first server was named, and null once the match is decided -- there is
   *  no next point, which is why scoreboardRows drops games and points
   *  there too. */
  server: Player | null
}

export const DEFAULT_RULES: ScoreRules = {
  players: ['Me', 'Opp'],
  sets: 3,
  ad: true,
  tiebreak: 'at6',
  tiebreakTo: 7,
  firstServer: null,
}

/**
 * Who serves the next point, from how many service units are done.
 *
 * Serve alternates every game, so the parity of completed games decides
 * it. A whole tiebreak counts as one unit, which is not a shortcut: the
 * player who would have served the next game serves the tiebreak's first
 * point, and treating the tiebreak as that game makes "whoever served
 * first in the tiebreak receives first in the next set" (ITF rule 5) fall
 * out of plain alternation instead of needing a special case.
 *
 * Inside a tiebreak the serve changes after the first point and every two
 * after that -- points 1 / 2,3 / 4,5 -- so the point about to be played,
 * 1-based, has had floor(p / 2) changes before it.
 */
function serverOf(
  rules: ScoreRules,
  units: number,
  inTb: boolean,
  pts: [number, number],
): Player | null {
  if (rules.firstServer === null) return null
  const flips = units + (inTb ? Math.floor((pts[0] + pts[1] + 1) / 2) : 0)
  return (['a', 'b'] as const)[(rules.firstServer === 'a' ? 0 : 1) ^ (flips % 2)]
}

const POINT_LABELS = ['0', '15', '30', '40']

function pointLabels(pts: [number, number], inTiebreak: boolean, ad: boolean): [string, string] {
  if (inTiebreak) return [String(pts[0]), String(pts[1])]
  const [a, b] = pts
  if (ad && a >= 3 && b >= 3) {
    if (a === b) return ['40', '40']
    return a > b ? ['Ad', '40'] : ['40', 'Ad']
  }
  return [POINT_LABELS[Math.min(a, 3)], POINT_LABELS[Math.min(b, 3)]]
}

export function score(winners: Player[], rules: ScoreRules): ScoreState {
  const setsNeeded = Math.floor(rules.sets / 2) + 1
  const sets: [number, number][] = []
  let games: [number, number] = [0, 0]
  let pts: [number, number] = [0, 0]
  let inTb = rules.tiebreak === 'only'
  let finished: Player | null = null
  let units = 0 // completed service units: games, plus a whole tiebreak as one
  const setsWonBy = (i: 0 | 1) => sets.filter((s) => s[i] > s[1 - i]).length

  for (const w of winners) {
    // Points after match point are ignored, not rejected: the queue shows
    // them as unscored, which is honest and needs no error state.
    if (finished) break
    const i: 0 | 1 = w === 'a' ? 0 : 1
    const j: 0 | 1 = i === 0 ? 1 : 0
    pts[i] += 1

    if (inTb) {
      if (pts[i] >= rules.tiebreakTo && pts[i] - pts[j] >= 2) {
        if (rules.tiebreak === 'only') {
          // One tiebreak is the whole session: the count is the final score.
          finished = w
        } else {
          games[i] += 1
          units += 1
          sets.push([games[0], games[1]])
          games = [0, 0]
          pts = [0, 0]
          inTb = false
          if (setsWonBy(i) >= setsNeeded) finished = w
        }
      }
      continue
    }

    // Four points and two clear with advantage scoring; without it the
    // seventh point of a game decides it at deuce.
    const wonGame = pts[i] >= 4 && (!rules.ad || pts[i] - pts[j] >= 2)
    if (!wonGame) continue
    games[i] += 1
    units += 1
    pts = [0, 0]
    if (rules.tiebreak === 'at6' && games[0] === 6 && games[1] === 6) {
      inTb = true
    } else if (games[i] >= 6 && games[i] - games[j] >= 2) {
      sets.push([games[0], games[1]])
      games = [0, 0]
      if (setsWonBy(i) >= setsNeeded) finished = w
    }
  }

  return {
    sets,
    games,
    points: pointLabels(pts, inTb, rules.ad),
    inTiebreak: inTb,
    finished,
    server: finished !== null ? null : serverOf(rules, units, inTb, pts),
  }
}

/**
 * The state entering `rallyId`, and how many earlier points carry no
 * winner. Replays every non-rejected point with a lower idx, in idx order
 * whatever order `rallies` arrived in -- the caller hands in the whole
 * session, not the source-scoped list the queue shows. An id no rally
 * holds replays everything.
 */
export function scoreBefore(
  rallies: Rally[],
  rallyId: string,
  rules: ScoreRules,
): { state: ScoreState; unscored: number } {
  const target = rallies.find((r) => r.id === rallyId)
  const earlier = rallies
    .filter((r) => (target ? r.idx < target.idx : true) && !r.rejected && r.point)
    .sort((x, y) => x.idx - y.idx)
  const winners = earlier.map((r) => r.winner).filter((w): w is Player => w !== '')
  const unscored = earlier.filter((r) => r.winner === '').length
  return { state: score(winners, rules), unscored }
}

export function playerName(rules: ScoreRules, p: Player): string {
  return rules.players[p === 'a' ? 0 : 1]
}

/** Two rows for a broadcast-style board: name, one column per completed
 *  set, current games, points. Tiebreak-only sessions have no games
 *  column; a finished match shows W in place of games and points. */
export function scoreboardRows(state: ScoreState, rules: ScoreRules): string[][] {
  return rules.players.map((name, i) => {
    const row = [name, ...state.sets.map((s) => String(s[i]))]
    if (state.finished !== null && rules.tiebreak !== 'only') {
      // Games and points are 0-0 after the deciding set; sets, then W.
      row.push(state.finished === (i === 0 ? 'a' : 'b') ? 'W' : '')
    } else {
      if (rules.tiebreak !== 'only') row.push(String(state.games[i]))
      row.push(state.points[i])
    }
    return row
  })
}

/**
 * One letter per player, for a control too small to hold a name.
 *
 * Initials, because "S" beside Sam's rally is read without translation;
 * positional A/B when they collide, because two players called Alex and
 * Ana rendering the same letter would be worse than no initials at all.
 * The fallback is all-or-nothing on purpose -- giving one player their
 * initial and the other a positional letter would read as one scheme with
 * a typo in it.
 */
export function playerGlyphs(rules: ScoreRules): [string, string] {
  const [a, b] = rules.players.map((n) => n.trim().charAt(0).toUpperCase())
  return a === b ? ['A', 'B'] : [a, b]
}

export interface WinnerChip {
  /** What the chip draws: the winner's letter, or an en dash for none. */
  glyph: string
  /** The same state in words, for the title attribute. */
  title: string
  /** Whether a winner is recorded -- the caller fills the chip on true. */
  set: boolean
}

/**
 * What QueueMode's fourth chip should say for a rally, or null when the
 * session tracks no score. Null rather than an empty chip: an always-there
 * winner box on an untracked session advertises a control that does
 * nothing.
 *
 * Takes the winner rather than the Rally because the queue controller's
 * live map is the one that reflects the row -- set_point(false) clears the
 * winner there and on the server, and the Rally the controller was built
 * from is deliberately never mutated.
 */
export function winnerChip(winner: Player | '', rules: ScoreRules | null): WinnerChip | null {
  if (!rules) return null
  const label = winnerLabel(winner, rules)
  if (label === null) return { glyph: '–', title: 'no winner recorded', set: false }
  return { glyph: playerGlyphs(rules)[winner === 'a' ? 0 : 1], title: label, set: true }
}

/** "won by Sam", or null when there is nothing to say -- no rules, or no
 *  winner recorded. Shared by the queue's chip and the overview band's
 *  label so the two cannot word the same fact differently. */
export function winnerLabel(winner: Player | '', rules: ScoreRules | null): string | null {
  if (!rules || winner === '') return null
  return `won by ${rules.players[winner === 'a' ? 0 : 1]}`
}
